File: src/cli/scrape-creator.php
<?php

use asset\ScrapedAsset;
use asset\StoredAssetQuery;
use creator\Creator;
use creator\CreatorLogic;
use log\LogLevel;
use database\Database;
use log\Log;
use log\LogResult;
use thumbnail\Thumbnail;
use misc\StringUtil;

require_once __DIR__ . '/../include/init.php';

// Show CLI syntax
if ($argc > 2 || in_array($argv[1] ?? '', ['-h', '--help', 'help', "h"])) {
	echo "Usage: php scrape-creator.php [creator-slug|creator-id] [force]
Not setting a creator will pick a random one from the regular scraping targets.
The force argument disables graceful backoff on errors.\n";
	exit(1);
}

// Start logging and determine the official run timestamp
$now = new DateTime();
$timestamp = $now->format('Y-m-d\TH-i-s-v');
Log::start(logName: "scrape-creator/"  . $timestamp, level: LogLevel::INFO, writeToStdout: true);

// Determine backoff behavior
$force = isset($argv[2]) && strtolower($argv[2]) === 'force';
Log::write("Forceful mode?", $force, LogLevel::DEBUG);

// Pick a target creator
if (isset($argv[1])) {
	$creator = Creator::fromAny($argv[1]);
} else {
	$creator = Creator::randomScrapingTarget(!$force);
}

/** 
 * @var Creator $creator
 * @var CreatorLogic $creatorLogic */
$creatorLogic = $creator->getLogic();
Log::write("Loaded logic and starting to scrape for creator", $creator->name, LogLevel::INFO);

// Increment failure counter (will be reset on success)
$creator->incrementFailedAttempts($now);

// Get existing assets to provide a comparison
$query = new StoredAssetQuery();
$query->filterCreator = [$creator];
$query->filterStatus = NULL;
$query->limit = NULL;
$existingAssets = $query->execute();
Log::write("Found " . sizeof($existingAssets) . " existing assets for creator.", LogLevel::INFO);

// Get new assets using creator-specific method
// Passing in the list of existing URLs and
$newScrapedAssets = $creatorLogic->scrapeAssets($existingAssets);

// Perform post-processing on the results
foreach ($newScrapedAssets as $scrapedAsset) {
	if ($scrapedAsset === null) {
		continue;
	}
	// Expand and clean up the tag array
	$titleWords = preg_split('/\s+/', $scrapedAsset->title);
	if ($titleWords !== false) {
		$scrapedAsset->tags = array_merge($scrapedAsset->tags, $titleWords);
	}
	$scrapedAsset->tags[] = $creator->slug();
	$scrapedAsset->tags = StringUtil::filterTagArray($scrapedAsset->tags);
}

Log::write("Found " . sizeof($newScrapedAssets) . " new assets", $newScrapedAssets, LogLevel::INFO);

// Save new assets to DB
if (sizeof($newScrapedAssets) > 0) {

	/**
	 * @var ScrapedAsset $newScrapedAsset
	 */
	foreach ($newScrapedAssets as $newScrapedAsset) {

		// Validity checks
		if ($newScrapedAsset->creator == null || $newScrapedAsset->creator !== $creator) {
			Log::write("Skipping asset with mismatched creator: ", $newScrapedAsset, LogLevel::WARNING);
			continue;
		}

		// Convert scraped asset to stored asset
		$newStoredAsset = $newScrapedAsset->toStoredAsset();

		// Save stored asset to DB
		Database::startTransaction();

		$newStoredAsset->writeToDatabase();
		if ($newStoredAsset->id !== null && $newScrapedAsset->rawThumbnailData !== null) {
			Thumbnail::saveThumbnailVariations($newStoredAsset->id, $newScrapedAsset->rawThumbnailData);
			Log::write("Saved thumbnail for asset ", $newStoredAsset->id);
		}

		Database::commitTransaction();

		Log::write("Committed new asset to DB", $newStoredAsset, LogLevel::INFO);
	}
	Log::write("Wrote " . sizeof($newScrapedAssets) . " new assets.", LogLevel::INFO);
} else {
	Log::write("No new updates to write to DB.", LogLevel::INFO);
}

$creator->resetFailedAttempts($now);

Thumbnail::deleteOrphanedThumbnails();

Log::stop(true);

File: cli/scrape_creator.py
#!/usr/bin/env python3
"""CLI: Scrape assets for a given creator."""
import sys
from pathlib import Path

# Ensure project root is on path when run directly
sys.path.insert(0, str(Path(__file__).parent.parent))

from dotenv import load_dotenv
load_dotenv()

from app.database.database import Database
from app.asset.stored_asset_query import StoredAssetQuery
from app.asset.stored_asset_status import StoredAssetStatus
from app.creator.creator import Creator
from app.log.log import Log
from app.misc.string_util import StringUtil
from app.thumbnail.thumbnail import Thumbnail


def main() -> None:
    args = sys.argv[1:]
    if not args:
        print("Usage: scrape_creator.py <creator-slug|creator-id> [force]")
        sys.exit(1)

    creator_arg = args[0]
    force = len(args) > 1 and args[1].lower() == "force"

    creator = Creator.from_value_or_slug(creator_arg)

    if not force and not creator.is_available_for_scrape():
        print(f"Creator {creator.slug()} is not available for scraping (too many failed attempts). Use 'force' to override.")
        sys.exit(0)

    Log.start(f"scrape-creator/{creator.slug()}")

    try:
        creator.increment_failed_attempts()

        # Load existing assets for this creator to check for duplicates
        existing_query = StoredAssetQuery(
            offset=0,
            limit=999999,
            filter_creator=[creator.slug()],
        )
        existing_assets = existing_query.execute()

        # Run the scraper
        logic = creator.get_logic()
        scraped_collection = logic.scrape_assets()

        saved_count = 0
        skipped_count = 0

        for scraped_asset in scraped_collection.assets:
            # Post-process tags
            scraped_asset.tags = StringUtil.filter_tag_array(scraped_asset.tags)
            # Add creator slug as a tag
            if creator.slug() not in scraped_asset.tags:
                scraped_asset.tags.append(creator.slug())

            # Skip if already exists
            if existing_assets.contains_url(scraped_asset.url):
                skipped_count += 1
                continue

            # Write to database
            stored = scraped_asset.to_stored_asset()
            Database.start_transaction()
            try:
                stored.write_to_database()
                Database.commit_transaction()
            except Exception:
                Database.rollback_transaction()
                raise

            # Save thumbnails
            if scraped_asset.raw_thumbnail is not None:
                try:
                    Thumbnail.save_thumbnail_variations(stored.id, scraped_asset.raw_thumbnail)
                except Exception as e:
                    Log.write(f"Failed to save thumbnail for asset {stored.id}: {e}")

            saved_count += 1

        creator.reset_failed_attempts()
        Thumbnail.delete_orphaned_thumbnails()
        Database.update_popularity_scores()

        Log.write(f"Scraped {len(scraped_collection.assets)} assets. Saved: {saved_count}, Skipped (already exists): {skipped_count}.")
    except Exception as e:
        Log.write(f"Error during scraping: {e}", error=True)
        raise
    finally:
        Log.stop()


if __name__ == "__main__":
    main()
